src/datanode/src/heartbeat.rs

```diff
@@ -103,10 +127,35 @@ impl HeartbeatTask {
         let meta_client = self.meta_client.clone();
 
         let catalog_manager_clone = self.catalog_manager.clone();
-        let mut tx = Self::create_streams(&meta_client, running.clone()).await?;
+        let mailbox = self.mailbox.clone();
+        let handler_executor = self.heartbeat_response_handler_exector.clone();
+        let mut tx = Self::create_streams(
+            &meta_client,
+            running.clone(),
+            handler_executor.clone(),
+            mailbox.clone(),
+        )
+        .await?;
+
         common_runtime::spawn_bg(async move {
             while running.load(Ordering::Acquire) {
                 let (region_num, region_stats) = datanode_stat(&catalog_manager_clone).await;
+                let mut mailbox_guard = mailbox.lock().await;
+
+                //TODO(weny): add a size-based limiter?
+                let mails = mailbox_guard.replies.drain(..);
+                let mailbox_messages = mails
+                    .into_iter()
+                    .map(outgoing_message_to_mailbox_message)
+                    .collect::<Result<Vec<MailboxMessage>>>();
+
+                if let Err(e) = mailbox_messages {
+                    error!(e;"Failed to encode mailbox messages!");
+                    continue;
```

Comment:
Even if there is an error in the mailbox, regular heartbeats still need to be sent.